pkg/controllers/provisioning/scheduling/nodeclaim.go

```diff
@@ -233,6 +241,8 @@ func filterInstanceTypesByRequirements(instanceTypes []*cloudprovider.InstanceTy
 		requirementsAndOffering: false,
 		fitsAndOffering:         false,
 	}
+	cumulativeMinRequirementsFromInstanceTypes := make(map[string]sets.Set[string])
```

Comment:
How expensive is it to rebuild this map every time vs. keeping a running total of key/values with counts and subtracting off based on an instance type's key if it's incompatible? 

Might be worth doing some benchmarking here with `testing.b`: https://gobyexample.com/testing-and-benchmarking to see which one is preferred here. From a Big-O complexity, I think they are pretty similar since you have to iterate through all the keys and then call `--` on each value that you find for the instance type but it's probably worth checking if the map instantiations combined with building new sets each time is expensive.

Reply:
Do not want to block rest of the PR. I will follow-up on this one separately since it is tricky to inject the `minvalues` into scheduler logic without API help for simulations.

Reply:
Here is the performance/benchmarking analysis for various scenarios:

**Without my changes**
```
scheduled 12610 against 2108 nodes in total in 22.999827136s 548.2649902295335 pods/sec
400 instances 10 pods    2 nodes    6.910374ms per scheduling    691.037µs per pod
400 instances 100 pods   17 nodes   74.308797ms per scheduling   743.087µs per pod
400 instances 500 pods   84 nodes   382.199278ms per scheduling  764.398µs per pod
400 instances 1000 pods  167 nodes  787.569292ms per scheduling  787.569µs per pod
400 instances 1500 pods  253 nodes  12.382227s per scheduling    8.254818ms per pod
400 instances 2000 pods  334 nodes  1.591589125s per scheduling  795.794µs per pod
400 instances 2500 pods  417 nodes  2.111523708s per scheduling  844.609µs per pod
400 instances 5000 pods  834 nodes  4.63844175s per scheduling   927.688µs per pod
PASS
ok  	sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling	29.812s
```

**With my changes and without minValues in requirement**
_( a slight increase which is expected due to additional loop to accumulate the requirements across InstanceTypeOptions which has `minValues` in it)_
```
scheduled 12610 against 2106 nodes in total in 23.767375254s 530.559216793523 pods/sec
400 instances 10 pods    2 nodes    7.280119ms per scheduling     728.011µs per pod
400 instances 100 pods   17 nodes   78.313497ms per scheduling    783.134µs per pod
400 instances 500 pods   84 nodes   407.551264ms per scheduling   815.102µs per pod
400 instances 1000 pods  167 nodes  812.94202ms per scheduling    812.942µs per pod
400 instances 1500 pods  251 nodes  12.858819459s per scheduling  8.572546ms per pod
400 instances 2000 pods  334 nodes  1.64878525s per scheduling    824.392µs per pod
400 instances 2500 pods  417 nodes  2.106910541s per scheduling   842.764µs per pod
400 instances 5000 pods  834 nodes  4.767597208s per scheduling   953.519µs per pod
PASS
ok  	sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling	30.899s
```

**With minValue = 3 for a requirement (I have taken InstanceType in all cases)**
_( little more increase in time to fetch the key that violates the `minValues` and this wouldn’t change no matter how much is the value of minValue as shown below for few other values)_
```
scheduled 12610 against 2105 nodes in total in 26.640151567s 473.345655270986 pods/sec
400 instances 10 pods    2 nodes    9.5229ms per scheduling       952.29µs per pod
400 instances 100 pods   17 nodes   100.883319ms per scheduling   1.008833ms per pod
400 instances 500 pods   84 nodes   506.510021ms per scheduling   1.01302ms per pod
400 instances 1000 pods  167 nodes  1.025684042s per scheduling   1.025684ms per pod
400 instances 1500 pods  251 nodes  13.503289875s per scheduling  9.002193ms per pod
400 instances 2000 pods  333 nodes  2.128573333s per scheduling   1.064286ms per pod
400 instances 2500 pods  417 nodes  2.723903459s per scheduling   1.089561ms per pod
400 instances 5000 pods  834 nodes  6.021298125s per scheduling   1.204259ms per pod
PASS
ok  	sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling	32.774s
```

**With minValue = 5 for a requirement**
```
scheduled 12610 against 2105 nodes in total in 26.527231966s 475.36056593323644 pods/sec
400 instances 10 pods    2 nodes    9.680388ms per scheduling     968.038µs per pod
400 instances 100 pods   17 nodes   103.119927ms per scheduling   1.031199ms per pod
400 instances 500 pods   84 nodes   506.872854ms per scheduling   1.013745ms per pod
400 instances 1000 pods  167 nodes  1.028840041s per scheduling   1.02884ms per pod
400 instances 1500 pods  251 nodes  13.454834917s per scheduling  8.969889ms per pod
400 instances 2000 pods  333 nodes  2.144536625s per scheduling   1.072268ms per pod
400 instances 2500 pods  417 nodes  2.686135083s per scheduling   1.074454ms per pod
400 instances 5000 pods  834 nodes  5.957702458s per scheduling   1.19154ms per pod
PASS
ok  	sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling	32.776s
```

**With minValue = 10 for a requirement**
```
scheduled 12610 against 2104 nodes in total in 26.174312078s 481.7700638099651 pods/sec
400 instances 10 pods    2 nodes    9.627454ms per scheduling    962.745µs per pod
400 instances 100 pods   17 nodes   100.234663ms per scheduling  1.002346ms per pod
400 instances 500 pods   83 nodes   513.745583ms per scheduling  1.027491ms per pod
400 instances 1000 pods  166 nodes  1.035767916s per scheduling  1.035767ms per pod
400 instances 1500 pods  251 nodes  13.0232s per scheduling      8.682133ms per pod
400 instances 2000 pods  334 nodes  2.174360208s per scheduling  1.08718ms per pod
400 instances 2500 pods  417 nodes  2.797213459s per scheduling  1.118885ms per pod
400 instances 5000 pods  834 nodes  5.968021292s per scheduling  1.193604ms per pod
PASS
ok  	sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling	31.468s
```

**With minValue = 15 for a requirement**
```
scheduled 12610 against 2104 nodes in total in 26.262151959s 480.1586716765064 pods/sec
400 instances 10 pods    2 nodes    9.596ms per scheduling        959.6µs per pod
400 instances 100 pods   17 nodes   100.269287ms per scheduling   1.002692ms per pod
400 instances 500 pods   83 nodes   509.614771ms per scheduling   1.019229ms per pod
400 instances 1000 pods  166 nodes  1.030548292s per scheduling   1.030548ms per pod
400 instances 1500 pods  251 nodes  13.134777458s per scheduling  8.756518ms per pod
400 instances 2000 pods  334 nodes  2.112007375s per scheduling   1.056003ms per pod
400 instances 2500 pods  417 nodes  2.796082833s per scheduling   1.118433ms per pod
400 instances 5000 pods  834 nodes  6.032787959s per scheduling   1.206557ms per pod
PASS
ok  	sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling	31.351s
```

**With minValue = 20 for a requirement**
```
scheduled 12610 against 2104 nodes in total in 26.592401554s 474.19560713211393 pods/sec
400 instances 10 pods    2 nodes    9.572666ms per scheduling     957.266µs per pod
400 instances 100 pods   17 nodes   98.358816ms per scheduling    983.588µs per pod
400 instances 500 pods   83 nodes   505.555083ms per scheduling   1.01111ms per pod
400 instances 1000 pods  166 nodes  1.035924208s per scheduling   1.035924ms per pod
400 instances 1500 pods  251 nodes  13.335883375s per scheduling  8.890588ms per pod
400 instances 2000 pods  334 nodes  2.157725417s per scheduling   1.078862ms per pod
400 instances 2500 pods  417 nodes  2.821344459s per scheduling   1.128537ms per pod
400 instances 5000 pods  834 nodes  6.098517833s per scheduling   1.219703ms per pod
PASS
ok  	sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling	31.832s
```

**With minValue = 30 for a requirement**
```
scheduled 12610 against 2104 nodes in total in 26.064046588s 483.80822054721534 pods/sec
400 instances 10 pods    2 nodes    9.512452ms per scheduling     951.245µs per pod
400 instances 100 pods   17 nodes   99.078468ms per scheduling    990.784µs per pod
400 instances 500 pods   83 nodes   506.811166ms per scheduling   1.013622ms per pod
400 instances 1000 pods  166 nodes  1.029381542s per scheduling   1.029381ms per pod
400 instances 1500 pods  251 nodes  13.102935959s per scheduling  8.73529ms per pod
400 instances 2000 pods  334 nodes  2.115130167s per scheduling   1.057565ms per pod
400 instances 2500 pods  417 nodes  2.698365667s per scheduling   1.079346ms per pod
400 instances 5000 pods  834 nodes  5.966055166s per scheduling   1.193211ms per pod
PASS
ok  	sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling	31.313s
```

**With minValue = 60 for a requirement**
```
scheduled 12610 against 2105 nodes in total in 26.556147719s 474.8429679421456 pods/sec
400 instances 10 pods    2 nodes    9.573855ms per scheduling     957.385µs per pod
400 instances 100 pods   17 nodes   100.061652ms per scheduling   1.000616ms per pod
400 instances 500 pods   84 nodes   504.558083ms per scheduling   1.009116ms per pod
400 instances 1000 pods  167 nodes  1.06675975s per scheduling    1.066759ms per pod
400 instances 1500 pods  251 nodes  13.458783708s per scheduling  8.972522ms per pod
400 instances 2000 pods  333 nodes  2.160281542s per scheduling   1.08014ms per pod
400 instances 2500 pods  417 nodes  2.699132708s per scheduling   1.079653ms per pod
400 instances 5000 pods  834 nodes  5.937499875s per scheduling   1.187499ms per pod
PASS
ok  	sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling	32.652s
```

**With minValue = 100 for a requirement (Cannot be a real usecase since the max is capped at 60 today to CreateFleet)**
```
scheduled 12610 against 2104 nodes in total in 26.30848654s 479.3130148641382 pods/sec
400 instances 10 pods    2 nodes    9.66196ms per scheduling      966.196µs per pod
400 instances 100 pods   17 nodes   99.390145ms per scheduling    993.901µs per pod
400 instances 500 pods   83 nodes   511.4355ms per scheduling     1.022871ms per pod
400 instances 1000 pods  166 nodes  1.141197958s per scheduling   1.141197ms per pod
400 instances 1500 pods  251 nodes  13.184101875s per scheduling  8.789401ms per pod
400 instances 2000 pods  334 nodes  2.158486542s per scheduling   1.079243ms per pod
400 instances 2500 pods  417 nodes  2.704895917s per scheduling   1.081958ms per pod
400 instances 5000 pods  834 nodes  5.958880209s per scheduling   1.191776ms per pod
PASS
ok  	sigs.k8s.io/karpenter/pkg/controllers/provisioning/scheduling	31.518s
```

_**There is an approx 1 second increase for the worst case scenario with 5000 pods and most other cases are in milliseconds increase in value. I think this is fine and expected. Let me know your thoughts.**_